### PR TITLE
B-Tree Index

### PR DESCRIPTION
**Pull request by ArneMayer**
Using Googles B-Tree implementation.
https://code.google.com/archive/p/cpp-btree/

### CODE REVIEW

File: src/lib/storage/index/b_tree/b_tree_index.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#include "b_tree_index.hpp"
+
+#include "storage/index/column_index_type.hpp"
+#include "resolve_type.hpp"
+
+namespace opossum {
+
+  BTreeIndex::BTreeIndex(const std::vector<std::shared_ptr<const BaseColumn>> index_columns)
+      : BaseIndex{get_index_type_of<BTreeIndex>()}, _index_column(index_columns[0]) {
+    DebugAssert((index_columns.size() == 1), "BTreeIndex only works with a single column.");
```
Comment (mrks) [inline src/lib/storage/index/b_tree/b_tree_index.cpp]: Let's make this a non-debug assert.

File: src/lib/storage/index/b_tree/b_tree_index_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+template <typename DataType>
+uint64_t BTreeIndexImpl<DataType>::memory_consumption() const {
+  return sizeof(std::vector<ChunkOffset>) +
+         sizeof(ChunkOffset) * _chunk_offsets.size() +
+         _btree.bytes_used();
+}
+
+template <typename DataType>
+void BTreeIndexImpl<DataType>::_bulk_insert(const std::shared_ptr<const BaseColumn> column) {
```
Comment (mrks) [inline src/lib/storage/index/b_tree/b_tree_index_impl.cpp]: Where is this tested?

**Comment by Bensk1** [comment]
Could you also please adapt DEPENDENCIES.md?

File: src/lib/storage/index/b_tree/b_tree_index_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  using Iterator = std::vector<ChunkOffset>::const_iterator;
+  virtual Iterator lower_bound(const std::vector<AllTypeVariant>&) const = 0;
+  virtual Iterator upper_bound(const std::vector<AllTypeVariant>&) const = 0;
+  virtual Iterator cbegin() const = 0;
+  virtual Iterator cend() const = 0;
+};
+
+/**
+* Implementation: https://code.google.com/archive/p/cpp-btree/
+* Note: does not support null values right now.
```
[the remark this answers is not in the document]
Comment (ArneMayer) [reply src/lib/storage/index/b_tree/b_tree_index_impl.hpp]: #944

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -68,6 +68,7 @@ include_directories(
     ${PROJECT_SOURCE_DIR}/third_party/cxxopts/include
     ${PROJECT_SOURCE_DIR}/third_party/json
     ${PROJECT_SOURCE_DIR}/third_party/sql-parser/src
+        ${PROJECT_SOURCE_DIR}/third_party/cpp-btree
```
[the remark this answers is not in the document]
Comment (ArneMayer) [reply src/CMakeLists.txt]: fixed.

File: src/lib/storage/index/b_tree/b_tree_index.cpp (hunk 1 above)
Comment (ArneMayer) [reply src/lib/storage/index/b_tree/b_tree_index.cpp]: Done and also changed this for the other indices.

**Comment by ArneMayer** [comment]
Updated DEPENDENCIES.md

File: src/lib/storage/index/b_tree/b_tree_index_impl.cpp (hunk 2 above)
Comment (ArneMayer) [reply src/lib/storage/index/b_tree/b_tree_index_impl.cpp]: I will write a test.

File: src/lib/storage/index/b_tree/b_tree_index_impl.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArneMayer) [reply src/lib/storage/index/b_tree/b_tree_index_impl.hpp]: We could just store all null value chunk offsets seperately. But then the question is how do the methods lower_bound and upper_bound handle null values? Are null values just always at the lower end of the value spectrum?

File: src/lib/storage/index/b_tree/b_tree_index_impl.cpp (hunk 2 above)
Comment (ArneMayer) [reply src/lib/storage/index/b_tree/b_tree_index_impl.cpp]: Done.

**Comment by mrks** [comment]
Could you add this to the table scan tests? `index_scan_test.cpp` and `join_index_test.cpp` already have a list of derived indexes.

Once that is done, we are good to go.

**Comment by ArneMayer** [comment]
Done.

**Pull request merged**